rust/phoenix-channel/src/heartbeat.rs

```diff
@@ -1,81 +1,93 @@
-use crate::{EgressControlMessage, OutboundRequestId};
+use crate::OutboundRequestId;
+use futures::FutureExt;
 use std::{
     pin::Pin,
+    sync::{atomic::AtomicU64, Arc},
     task::{ready, Context, Poll},
     time::Duration,
 };
 use tokio::time::MissedTickBehavior;
 
-const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(30);
+pub const INTERVAL: Duration = Duration::from_secs(30);
+pub const TIMEOUT: Duration = Duration::from_secs(5);
 
 pub struct Heartbeat {
     /// When to send the next heartbeat.
     interval: Pin<Box<tokio::time::Interval>>,
+
+    timeout: Duration,
+
     /// The ID of our heatbeat if we haven't received a reply yet.
-    id: Option<OutboundRequestId>,
+    pending: Option<(OutboundRequestId, Pin<Box<tokio::time::Sleep>>)>,
+
+    next_request_id: Arc<AtomicU64>,
 }
 
 impl Heartbeat {
-    pub fn maybe_handle_reply(&mut self, id: OutboundRequestId) -> bool {
-        let Some(pending) = self.id.take() else {
-            return false;
-        };
+    pub fn new(interval: Duration, timeout: Duration, next_request_id: Arc<AtomicU64>) -> Self {
+        let mut interval = tokio::time::interval(interval);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
 
-        if pending != id {
-            return false;
+        Self {
+            interval: Box::pin(interval),
+            pending: Default::default(),
+            next_request_id,
+            timeout,
         }
-
-        self.id = None;
-        true
     }
 
-    pub fn set_id(&mut self, id: OutboundRequestId) {
-        self.id = Some(id);
+    pub fn maybe_handle_reply(&mut self, id: OutboundRequestId) -> bool {
+        match self.pending.as_ref() {
+            Some((pending, timeout)) if pending == &id && !dbg!(timeout.is_elapsed()) => {
+                self.pending = None;
+
+                true
+            }
+            _ => false,
+        }
     }
 
     pub fn poll(
         &mut self,
         cx: &mut Context,
-    ) -> Poll<Result<EgressControlMessage<()>, MissedLastHeartbeat>> {
-        ready!(self.interval.poll_tick(cx));
-
-        if self.id.is_some() {
-            self.id = None;
+    ) -> Poll<Result<OutboundRequestId, MissedLastHeartbeat>> {
+        if let Some((_, timeout)) = self.pending.as_mut() {
+            ready!(timeout.poll_unpin(cx));
+            self.pending = None;
             return Poll::Ready(Err(MissedLastHeartbeat {}));
         }
 
-        Poll::Ready(Ok(EgressControlMessage::Heartbeat(crate::Empty {})))
-    }
+        ready!(self.interval.poll_tick(cx));
 
-    fn new(interval: Duration) -> Self {
-        let mut interval = tokio::time::interval(interval);
-        interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+        let next_id = self
+            .next_request_id
+            .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
```

Comment:
Why have an atomic value here? Since you already have `&mut self`

Reply:
https://github.com/firezone/firezone/issues/4317